### PR TITLE
feat(region): expose ownership reads

### PR DESCRIPTION
**Pull request by paulfalgout**
## Linked issue

- Related #132

## What

- Add pure `Region#getOwner()` and `Region#getName()` queries over the existing View registration.
- Reject conflicting Region ownership, occupied names, destroyed instances, duplicate batches, and hook-induced reentrancy with stable diagnostic `MN0030`.
- Preserve empty-string Region names and unnamed CollectionView internal cleanup; document and test ordered partial application when a later `addRegions` entry fails.

## Why

Region ownership already existed privately, but consumers and coding agents could not inspect it without relying on `_parentView` and `_name`. The v3/v4 registration path also silently overwrote ownership and name records, allowing two registries to claim one Region or leaving an occupied name's former Region orphaned.

This makes the stored relationship explicit and authoritative. It deliberately rejects ambiguous mutation instead of introducing implicit transfer or replacement semantics that would also need to redefine parent element scope, child-event proxy ownership, monitoring, DOM handling, and lifecycle order.

## Scope

- Production Region/View registration and teardown paths.
- Diagnostic catalog, migration/reference documentation, lifecycle and reentrancy tests, changelog, and generated distributions.
- Intentionally does not add Region transfer, implicit replacement, child lookup aliases, or Application ownership APIs.
- Verified consumer impact: app-frontend currently adds a `content` Region in `onRender`; it must move that Region to a stable declaration and only show its child during render. marionette.toolkit has no View `addRegion` usage; Application `setRegion` semantics remain deferred to #190.

## Deployment Impact

- No website, npm package, tag, release, or deployment is performed by this PR.
- Consumers adopting this commit must migrate repeated same-name `addRegion` calls before exercising rerender paths.

## Agent-development failure addressed

- Private ownership fields were the only way to answer who owns a Region and under which name.
- Silent overwrite behavior made topology non-authoritative and difficult for humans or agents to diagnose.
- Reentrant lifecycle hooks and ordered batches now have explicit, tested outcomes instead of emergent registry corruption.

## Public behavior contract

- Canonical behavior after this PR: `getOwner()` and `getName()` are side-effect-free; one live Region has at most one registered owner/name; conflicts throw `MN0030` before the conflicting registration commits. A later ordered batch failure preserves earlier successful registrations and removes conflicting/later declarations.
- Superseded behavior removed, if any: silent Region instance reuse and occupied-name overwrite are no longer supported.
- Compatibility exception and removal condition, if any: none. Active consumers must use stable Region declarations or remove and replace explicitly with a fresh Region.

## Runtime cost

- [ ] Static or documentation only
- [ ] Development or test only; excluded from production entrypoints
- [x] Changes an existing production path
- [ ] Adds an explicitly opt-in runtime path

Measured impact or explanation:

- Bundle: cumulative measured output remained within budget at 64.06 kB / 75.00 kB.
- Startup/hot path: no startup work; registration adds bounded own-property and lifecycle-state checks. Timing benchmarks completed without a reported budget failure.
- Allocations/retention: ownership reads allocate nothing; multi-Region preflight uses one short-lived `Set` per `addRegions` call and no second ownership registry.

## Validation

- [x] Tests cover the acceptance criteria and relevant edge cases.
- [x] Docs, examples, declarations, diagnostics, and rule metadata agree where relevant.
- [x] No private consumer, private fixture, or undocumented context is required.
- [x] I ran the commands listed below and am reporting their actual results.

Commands:

```sh
npm test -- --run
# 94 files, 1735 tests passed

npm test -- --run test/unit/view-region-diagnostics.spec.js test/unit/region-lifecycle.spec.js test/unit/collection-view/collection-view-empty.spec.js
# 3 files, 123 tests passed

npm run lint:base -- --quiet
npm run docs:check
# 17 executable examples; 43 HTML files; 394 internal links

npm run check:diagnostics
# 30 catalog entries validated

npm run size
# 64.06 kB / 75.00 kB cumulative

npm run performance:timing
npm run check:dist
```

Exact-head bounded Claude review: approved with no blockers for `d097ef9148e752709010b9e2f6df6a13f9ed669d`; artifact `/Users/paulfalgout/.ai-reviews/20260831T050419Z-review-with-claude.md`.

## Package and browser impact

- Entry points or install fixtures affected: main package exports inherit the new Region prototype methods; generated main distributions are updated. No new entrypoint or dependency.
- Browser contracts affected: Region ownership and registration behavior only; no browser-support target or DomApi contract changes.
- Production/dev/test module-graph impact: existing `modules/view-region.js` path only; no new module edge.

## Rollback or deprecation

- Revert this commit to restore private-only ownership reads and silent overwrite behavior. No data migration or persisted state is involved.

## Review notes

- v4.1.3 source was inspected: `_addRegion` overwrote `_parentView`, `_name`, and `_regions[name]`; `removeRegion` destroyed rather than transferred a Region.
- Implicit transfer was rejected because `parentEl`, current child DOM, event proxy ownership, monitoring, and lifecycle ordering would all require a larger public contract.
- Same-View and cross-View hook reentrancy, duplicate batches, invalid later definitions, empty-string names, direct removal, destroying/destroyed instances, and the CollectionView unnamed internal Region are covered explicitly.

---
## Summary by cubic
Adds `Region#getOwner()` and `Region#getName()` as side-effect-free ownership queries and replaces silent region registration overwrites with `MN0030` conflict errors.

- `getOwner()` and `getName()` read the existing owner/name records without rendering or changing topology.
- Re-adding a Region under its current owner and name is an idempotent no-op that fires no lifecycle events.
- Registration now throws `MN0030` for already-owned, destroying, or destroyed regions; occupied names; and duplicate instances in a single batch.
- Reentrant conflicts from `before:add:region` hooks and later batch failures keep earlier registrations and drop conflicting and later entries.
- Empty-string region names and unnamed CollectionView internal region teardown still clear ownership correctly.

**Migration**

- Consumers that registered the same region name repeatedly must declare regions stably or remove and replace with a fresh region instance.

<sup>Written for commit 16cbea507d4b559d721589e4dcbcd3305387447a. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/marionettejs/marionette/pull/324?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

## Summary by CodeRabbit

- **New Features**
  - Added `getOwner()` and `getName()` methods for checking a region’s current owner and registered name.
  - Region ownership and naming now update correctly during registration, removal, and destruction.
- **Bug Fixes**
  - Added clear diagnostics for duplicate regions, conflicting names, invalid ownership, and destroyed regions.
  - Improved cleanup and consistency when registration fails or changes occur during registration.
  - Preserved successful earlier registrations when a later batch registration conflicts.
- **Documentation**
  - Updated migration guidance with the new Region ownership and naming behavior.

### CODE REVIEW

**Comment by paulfalgout** [comment]
**Approved on the code** at `d097ef91` (the SHA I fetched). **Bundle size needs a maintainer growth approval** targeting that exact head — +1,042 B, +1.65% aggregate (64,026 B against the 75,000 B ceiling, 10,974 B headroom), with all four main artifacts between +1.56% and +1.75%. Nothing is over budget; it is purely the >1% gate.

Incidentally, this is the first PR where #322's work actually runs — the report now carries a **Canonical consumer bundles** section, reporting-only as designed.

## The `_name` change is a pre-existing bug fix, not feature scaffolding

```diff
-    if (this._name) {
+    if (this._parentView && this._name !== undefined) {
```

A Region registered under the empty string failed that truthiness check, so `_removeReferences('')` never ran on destroy. The destroyed Region stayed in its owner's `_regions`, and `hasRegion('')` would go on reporting a dead Region indefinitely. Same shape as the `Radio.reset` arity guard from #211 — a falsy-but-valid value failing a truthiness test — which is now the fourth distinct instance of that family in this milestone.

It is properly pinned: `clears ownership for a Region registered under an empty name` asserts `getName() === ''` while registered and `hasRegion('') === false` afterward, so the assertion would fail if the guard regressed to truthiness. The CollectionView empty-Region case is covered separately, where `_parentView` is set but `_name` is undefined and `_removeReferences` must be skipped while ownership is still cleared.

## The doubled `assertRegionCanRegister` is the part worth defending

Calling it before `triggerMethod('before:add:region')` **and** again after is not belt-and-braces. A hook can adopt the Region or claim the name during that dispatch, so a single pre-check is stale by the time `region._parentView = this` executes. That window only opens under a listener nobody has written yet, which is exactly the kind that ships.

The catch there is equally careful: `delete this.regions[name]` runs only when no Region actually occupies that name, so a definition installed by the reentrant hook survives rather than being cleaned up by the losing caller.

Related and easy to read as cosmetic: swapping `_addRegion` ahead of `setRegion` is load-bearing. Registering before recording means a throw leaves the returned map without the failed entry; the old order would hand the caller a Region that was never registered.

## Ordered partial application deserves a caller-facing sentence

`addRegions({ a, b, c })` where `b` fails leaves `a` registered and `c` untouched. That is defensible — the batch is a convenience over a per-entry API, and the pre-flight `assertRegionDefinitionsCanRegister` catches the predictable conflicts before any mutation — but it is a trap for anyone who reads a batch call as atomic. The rollback removes only *definitions* for names that never registered; it deliberately does not unwind successful registrations.

Since the PR already documents and tests it, the remaining gap is just that a reader encountering `addRegions` in the reference would not learn it there. Worth one line wherever `addRegions` is documented, not a code change.

Nothing blocking on the code; it needs the growth approval to merge.

**Comment by paulfalgout** [comment]
```json
{
  "schemaVersion": 1,
  "headSha": "d097ef9148e752709010b9e2f6df6a13f9ed669d",
  "issueUrl": "https://github.com/marionettejs/marionette/issues/127",
  "approvedPaths": [
    "dist/marionette.cjs",
    "dist/marionette.js",
    "dist/marionette.min.js",
    "dist/marionette.umd.js"
  ],
  "evidenceUrls": [
    "https://github.com/marionettejs/marionette/issues/127#issuecomment-5474008658"
  ]
}
```

File: docs/migration-from-v4.md [hunk 1]
```diff
@@ -86,6 +86,7 @@ Status values describe the v5 outcome:
...
+| Region owner and name queries | Region ownership was maintained privately as `_parentView` and `_name`; reusing a Region instance or occupied name could leave conflicting private ownership records. | `Region#getOwner()` and `Region#getName()` expose the one current registered relationship without rendering, resolving elements, allocating a second registry, or mutating topology. Standalone and successfully destroyed Regions return `undefined`. Registration rejects already owned, destroying, or destroyed Regions and occupied names with `MN0030` before committing the conflicting registration. If a lifecycle hook creates a conflict during ordered `addRegions` processing, earlier entries remain registered while the conflicting and later entries do not. | Added | Use these methods instead of reading private fields. Remove an existing named Region before replacing it, and use a fresh Region instance for another owner. For render-time setup, declare a stable Region in `regions` and only show its child from `onRender`. Keep `currentView` and `hasView()` for the Region's owned child rather than adding a second child lookup API. |
```
[the remark this answers is not in the document]
Comment (paulfalgout) [reply docs/migration-from-v4.md]: Not changing this because #132 explicitly requires minimal public read-only ownership accessors, and ROADMAP.md makes public read-only ownership/topology a v5 core contract.

**Comment by paulfalgout** [comment]
**Approval stands** at `16cbea50` — re-reviewed rather than carried over, since this push changes runtime. The growth approval now needs to target this head: **+1,240 B, +1.97% aggregate** (64,224 B against the 75,000 B ceiling, 10,776 B headroom).

## Idempotent re-registration is the right carve-out

The first version turned a silent overwrite into a hard `MN0030` even for the harmless case of re-declaring a registration that already exists. Exempting it is correct, and the three-clause identity check is what makes the exemption safe:

```js
region._parentView === view && region._name === name &&
  getOwnRegion(view._regions, name) === region
```

The third clause carries the weight. The first two only establish what the Region *believes* about itself; the registry lookup confirms the owner agrees. A Region left holding stale `_parentView` / `_name` after a partial failure would satisfy the first two and be correctly rejected by the third — which matters precisely because this PR introduces new partial-failure paths.

## The no-op is event-silent, and that is pinned

Returning before `triggerMethod('before:add:region')` means a repeat registration fires no hook — the correct reading of "no mutation, no event." Two tests pin it rather than leaving it implicit: `keeps mixed batch events isolated from an identical registration no-op` and `keeps a reentrant identical registration authoritative without a duplicate add event`.

Those earn their place. Moving the early return below the hook would still produce correct *state* and break nothing except event counts — so without assertions on the events, that regression would pass a state-only suite.

## The destroying-window exemption checks out

`isSameRegionRegistration` short-circuits ahead of the destroying rejection, so re-adding an identical registration mid-teardown returns instead of throwing. Safe, and for a reason worth stating: identity match *requires* the Region to already occupy that name, so there is nothing to mutate. Once destruction completes, `_removeReferences` deregisters it, the third clause stops matching, and a later call lands on the destroyed-Region rejection as intended. The window is narrow and self-closing.

## The batch restructure was necessary, not tidying

Moving the name-collision check behind the `instanceof Region` branch is what lets the identity carve-out reach batch entries at all — under the old ordering, a repeated Region instance would have been rejected by the collision check before `assertRegionCanRegister` ever ran. A plain definition on an occupied name still throws, which is right: it would construct a *new* Region and genuinely conflict. `rejects a plain definition for an occupied batch name before mutation` covers that half.

**Comment by paulfalgout** [comment]
```json
{
  "schemaVersion": 1,
  "headSha": "16cbea507d4b559d721589e4dcbcd3305387447a",
  "issueUrl": "https://github.com/marionettejs/marionette/issues/127",
  "approvedPaths": [
    "dist/marionette.cjs",
    "dist/marionette.js",
    "dist/marionette.min.js",
    "dist/marionette.umd.js"
  ],
  "evidenceUrls": [
    "https://github.com/marionettejs/marionette/issues/127#issuecomment-5474219127"
  ]
}
```

**Comment by paulfalgout** [comment]
@cubic-dev-ai please re-review the exact current head 16cbea507d4b559d721589e4dcbcd3305387447a. The existing check has remained in progress since 05:37:33Z after all other gates completed.
